P6cl-99: Goal_buddies -> Get All

f2xiao opened the pull request:
This PR add a firebase function to be able to fetch all goal buddies data from firebase. The main changes include:

✅ add a file called goalBuddies.ts under src/firebase/functions directory
Moved utils functions from functions/ to a new utils/ directory.

✅add a function called getAllGoalBuddies

**How To Test**

1. replace all code in src/pages/ColabPage/ColabPage.tsx with the following code
```
import { useEffect } from 'react'
import { getAllGoalBuddies } from '../../firebase/functions/goalBuddies'

export default function ColabPage() {
  useEffect(() => {
    const fetchAllGoalBuddies = async () => {
      const data = await getAllGoalBuddies()
      console.log(data)
    }

    fetchAllGoalBuddies()
  }, [])

  return (
    <main>
      <h1>Welcome Colab User</h1>
    </main>
  )
}
```
2. inspect browser Console in the developer mode, and it should look like the following image
![image](https://github.com/user-attachments/assets/e3e1c462-2128-402d-b506-c6eec65071b6)

robester0403 commented:
There is also a Generics way of defining this function. I feel like with noSQL and the idea of typescript is to make sure we know what's in the data. So I would advise maybe not using Generics here. 

The counter argument is that using a noSQL means you SHOULD be flexible in adding any key data pair. If that was the case for choosing noSQL then I would use generics instead of explicit typing

KevinMeisenbacher commented:
![image](https://github.com/user-attachments/assets/a2f62516-2b95-480c-89bd-4d5b70a40cc1)
I downloaded the latest commit and put this in my .env. If this is working for everyone else, then any ideas on making it work right for me?

f2xiao commented on src/firebase/functions/goalBuddies.ts:
```diff
@@ -0,0 +1,21 @@
+    const querySnapshot = await getDocs(collection(db, 'goal_buddies'))
+    const goalBuddiesArray = querySnapshot.docs.map((doc) => ({
+      id: doc.id,
+      ...doc.data(),
```
thanks for the suggestion, Robert! 

I am adding the BaseGoalBuddy interface, however, `...doc.data()` doesn't automatically infer types or include `id` field. 

My solution: 
- having two interface `BaseGoalBuddy` (w/o id) and `GoalBuddy` (w/ id)
- explicitly infer the `...doc.data() as BaseGoalBuddy` to elimit error warning
- include a `GoalBuddy` interface to extend `BaseGoalBuddy` with the `id` field

![image](https://github.com/user-attachments/assets/f0a33908-a478-4715-a8c1-e159f3b127fd)
![image](https://github.com/user-attachments/assets/154f3591-98a8-40c6-849b-c3c4d437d68f)
![image](https://github.com/user-attachments/assets/f2978707-597d-4724-bd1a-5daa9fd8d822)


any suggestion other than this?

f2xiao commented:
> > > As well, please rename the file. I believe as per our lunch and learn we were going to have a file for each function. Rather than all goal_buddies collection functions in one file. so if this is the "getAllGoalBuddies" function, then that should be the name of the file.
> > > Tested and met the success criteria (so exciting so see it coming together!)
> > 
> > 
> > Having one file per collection is more organized and modular, makes it easier to manage and scale.
> 
> Honestly, I agree. We just need to make a note of it so that we consistently do this as a team.
> 
> Robert had some good feedback, but understand if under the time constraint you do not feel it can implemented in time for this Sprint. Could you help make sure (I will try to keep record of it too, but two heads better than one) that a ticket gets written for the next Sprint so that this feedback gets implemented going forward?

I have addressed the feedback given by Robert and waiting for his reply. Given the time constraint I do agree this issue is better to go to the next sprint.

f2xiao commented:
> Also please make sure that you have pulled and are updated. When I look at the README file when testing your branch, I do not see the changes that were merged into develop from Jeffrey's PR.

My ticket is not about READ.me, once my ticket is merged to `develop`, it will add `goBuddies.ts` and `types.ts` , the `develop` will still have Jeffery's updates on READ.me

AD9-1 commented:
Hello @f2xiao, You created the types.ts file in the utils folder, since it is a Firebase-related interface, it is better to move the folder under the firebase folder.

f2xiao commented:
We are not working on the same files, there is no conflicts. Even if we are
working on the same files, the conflict should be resolved while you merge
to develop. My next new ticket will branch out from the latest develop
which will have the new changes. If you still have confusion, we can have a
team meeting to clarify the git workflow with Robert.

On Mon, Feb 3, 2025, 12:39 PM Katrina Kite ***@***.***> wrote:

> Yes, but if you have pulled from develop, it should have the updated
> README from Jeffrey's merged PR request yes?
> If not it would get overwritten and the change lost?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/makeitMVPadmin/LAP6_colab/pull/11#issuecomment-2631651097>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGAGJQPSZLXAIZFXDH2M7VD2N6STHAVCNFSM6AAAAABWK6CFFOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMZRGY2TCMBZG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

f2xiao commented:
This will also be used on the frontend components.

On Mon, Feb 3, 2025, 12:45 PM Aparna Dhara ***@***.***> wrote:

> Hello @f2xiao <https://github.com/f2xiao>, You created the types.ts file
> in the utils folder, since it is a Firebase-related interface, it is better
> to move the folder under the firebase folder.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/makeitMVPadmin/LAP6_colab/pull/11#issuecomment-2631665615>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGAGJQPNUQTP4GTUBUPYTBL2N6TNFAVCNFSM6AAAAABWK6CFFOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMZRGY3DKNRRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

f2xiao commented:
> Please make sure to update from develop. Another PR was approved and merged while this PR was waiting for final confirmation.

Done

f2xiao commented:
> Hello @f2xiao, You created the types.ts file in the utils folder, since it is a Firebase-related interface, it is better to move the folder under the firebase folder.

This will also be used on the frontend components.
